Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1297 @@
+# Parsing with Haskell (Part 2): Parsing with Happy
+
+<!--
+* [X] refer to LR(1) article
+* [X] link to its user guide
+-->
+
+This is the second and final part of the _Parsing with Haskell_ series.
+Haven't read the first part yet? You can find it [here](https://serokell.io/blog/lexing-with-alex).
+
+As with Alex, our goal here is to provide enough information on Happy enough to be productive and do valuable things with it so that we won't visit every single feature of Happy.
+Please refer to the [Happy User Guide](https://monlih.github.io/happy-docs/).
+
+This tutorial was written using GHC version 9.0.2, Stack resolver LTS-19.8, Alex version 3.2.7.1, and Happy version 1.20.0.
+
+## Our grammar: MiniML
+
+We've discussed the grammar in the previous article, but let's reintroduce it here.
+MiniML (read as "minimal") will be a small grammar based on ML.
+By the end of this tutorial, you will be able to parse programs such as the following:
+
+```sml
+let the_answer : int =
+  let a = 20 in
+  let b = 1 in
+  let c = 2 in
+  a * c + b * c
+
+let main (unit : ()) : () =
+  print ("The answer is: " + the_answer)
+```
+
+Some of the features that this tutorial will support are:
+* Variables, constants, function declarations, function applications, local bindings, and operators.
+* Some simple types, as well as arrow types (functions).
+* Conditional expressions: if-then-else and if-then.
+* Block comments (added in part 1).
+* The reader will be invited in the proposed exercises to extend the language with patterns, list access, and pattern matching. The exercises in the first part also introduced rational numbers and line comments.
+
+## Comparison with parser combinators
+
+<!--
+* [X] refer to parser combinators article
+* [X] talk about the pros and cons
+-->
+
+In a previous tutorial, we discussed implementing and using [Parser Combinators in Haskell](https://serokell.io/blog/parser-combinators-in-haskell).
+In the article, we discussed some pros and cons of LL parsers compared to LR parsers, which we will reiterate and elaborate on.
+
+<!--
+TODO: not sure how correct is the theory here, a more in-depth review here would be nice
+-->
+
+Haskell libraries such as Megaparsec are typically exposed as a LL(1) or LL(∞) parser.
+LL here means **L**eft-to-right parsing, **L**eftmost derivation, as they scan strings from left to right and build a tree deriving them from the left (see below).
+
+The number between the parentheses indicates how many look-ahead tokens the parser can see.
+A parser with a look-ahead of `k` can make decisions by looking `k` tokens (or characters) ahead.
+A look-ahead of `k=1` is often better for better performance, and it's enough to parse most grammars in programming languages.
```

Comment:
Thanks for the correction, in Appel's book, there is this figure:

![image](https://user-images.githubusercontent.com/30162364/171913489-944ff40b-d8e2-4bad-b592-0542686863f0.png)

I was under the impression that LL(1) is pretty powerful due to the size of its bubble. I changed it to LL(k) and I hope it will be more adequate than saying LL(1) is enough. Please let me know if it's still not adequate.